File: app/src/main/java/com/limelight/binding/input/driver/AbstractDualSenseController.java
package com.limelight.binding.input.driver;

import android.hardware.usb.UsbConstants;
import android.hardware.usb.UsbDevice;
import android.hardware.usb.UsbDeviceConnection;
import android.hardware.usb.UsbEndpoint;
import android.hardware.usb.UsbInterface;
import android.os.SystemClock;
import android.util.Log;

import com.limelight.nvstream.input.ControllerPacket;
import com.limelight.nvstream.jni.MoonBridge;

import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.util.ArrayList;
import java.util.List;

public abstract class AbstractDualSenseController extends AbstractController {
    protected final UsbDevice device;
    protected final UsbDeviceConnection connection;

    private Thread inputThread;
    private boolean stopped;

    protected UsbEndpoint inEndpt, outEndpt;

    // IMU data fields
    protected float gyroX, gyroY, gyroZ;
    protected float accelX, accelY, accelZ;

    public AbstractDualSenseController(UsbDevice device, UsbDeviceConnection connection, int deviceId, UsbDriverListener listener) {
        super(deviceId, listener, device.getVendorId(), device.getProductId());
        this.device = device;
        this.connection = connection;
        this.type = MoonBridge.LI_CTYPE_PS;
        this.capabilities = MoonBridge.LI_CCAP_GYRO | MoonBridge.LI_CCAP_ACCEL | MoonBridge.LI_CCAP_RUMBLE;
        this.buttonFlags =
                ControllerPacket.A_FLAG | ControllerPacket.B_FLAG | ControllerPacket.X_FLAG | ControllerPacket.Y_FLAG |
                        ControllerPacket.UP_FLAG | ControllerPacket.DOWN_FLAG | ControllerPacket.LEFT_FLAG | ControllerPacket.RIGHT_FLAG |
                        ControllerPacket.LB_FLAG | ControllerPacket.RB_FLAG |
                        ControllerPacket.LS_CLK_FLAG | ControllerPacket.RS_CLK_FLAG |
                        ControllerPacket.BACK_FLAG | ControllerPacket.PLAY_FLAG | ControllerPacket.SPECIAL_BUTTON_FLAG;
        this.supportedButtonFlags = this.buttonFlags;
    }

    private Thread createInputThread() {
        return new Thread() {
            public void run() {
                try {
                    // Delay for a moment before reporting the new gamepad and
                    // accepting new input. This allows time for the old InputDevice
                    // to go away before we reclaim its spot. If the old device is still
                    // around when we call notifyDeviceAdded(), we won't be able to claim
                    // the controller number used by the original InputDevice.
                    Thread.sleep(1000);
                } catch (InterruptedException e) {
                    return;
                }

                // Report that we're added _before_ reporting input
                notifyDeviceAdded();

                while (!isInterrupted() && !stopped) {
                    byte[] buffer = new byte[64];

                    int res = -1; // Initialize to error state

                    //
                    // There's no way that I can tell to determine if a device has failed
                    // or if the timeout has simply expired. We'll check how long the transfer
                    // took to fail and assume the device failed if it happened before the timeout
                    // expired.
                    //

                    do {
                        // Check if we should stop before attempting transfer
                        if (stopped || isInterrupted()) {
                            res = -1; // Set to error state before break
                            break;
                        }

                        // Read the next input state packet
                        long lastMillis = SystemClock.uptimeMillis();
                        if (connection == null || inEndpt == null) {
                            Log.w("DualSenseController", "Connection or endpoint is null");
                            res = -1; // Set to error state before break
                            break;
                        }
                        res = connection.bulkTransfer(inEndpt, buffer, buffer.length, 3000);

                        // If we get a zero length response, treat it as an error
                        if (res == 0) {
                            res = -1;
                        }

                        if (res == -1 && SystemClock.uptimeMillis() - lastMillis < 1000) {
                            Log.d("DualSenseController", "Detected device I/O error");
                            AbstractDualSenseController.this.stop();
                            break;
                        }
                    } while (res == -1 && !isInterrupted() && !stopped);

                    if (res == -1 || stopped || isInterrupted()) {
                        break;
                    }

                    if (res > 0 && handleRead(ByteBuffer.wrap(buffer, 0, res).order(ByteOrder.LITTLE_ENDIAN))) {
                        // Report input if handleRead() returns true
                        reportInput();
                        reportMotion();
                    }
                }
            }
        };
    }

    private static UsbInterface findInterface(UsbDevice device) {
        int count = device.getInterfaceCount();
        for (int i = 0; i < count; i++) {
            UsbInterface intf = device.getInterface(i);
            if (intf.getInterfaceClass() == UsbConstants.USB_CLASS_HID && intf.getEndpointCount() >= 2) {
                Log.d("DualSenseController", "Found HID interface: " + i);
                return intf;
            }
        }
        return null;
    }

    private List<UsbInterface> ifaces = new ArrayList<>();

    public boolean start() {
        ifaces.clear();
        Log.d("DualSenseController", "start");
        // Force claim all interfaces
        for (int i = 0; i < device.getInterfaceCount(); i++) {
            UsbInterface iface = device.getInterface(i);

            if (!connection.claimInterface(iface, true)) {
                Log.d("DualSenseController", "Failed to claim interface: " + i);
                return false;
            } else {
                ifaces.add(iface);
            }
        }
        Log.d("DualSenseController", "getInterfaceCount:" + device.getInterfaceCount());

        // Find the endpoints
        UsbInterface iface = findInterface(device);

        if (iface == null) {
            Log.e("DualSenseController", "Failed to find interface");
            return false;
        }

        for (int i = 0; i < iface.getEndpointCount(); i++) {
            UsbEndpoint endpt = iface.getEndpoint(i);
            if (endpt.getDirection() == UsbConstants.USB_DIR_OUT) {
                if (outEndpt != null) {
                    Log.d("DualSenseController", "Found duplicate OUT endpoint");
                    return false;
                }
                outEndpt = endpt;
            } else if (endpt.getDirection() == UsbConstants.USB_DIR_IN) {
                if (inEndpt != null) {
                    Log.d("DualSenseController", "Found duplicate IN endpoint");
                    return false;
                }
                inEndpt = endpt;
            }
        }
        Log.d("DualSenseController", "inEndpt: " + inEndpt);
        Log.d("DualSenseController", "outEndpt: " + outEndpt);
        // Make sure the required endpoints were present
        if (inEndpt == null || outEndpt == null) {
            Log.d("DualSenseController", "Missing required endpoint");
            return false;
        }
        // Run the init function
        if (!doInit()) {
            return false;
        }
        // Start listening for controller input
        inputThread = createInputThread();
        inputThread.start();
        return true;
    }

    public void stop() {
        synchronized (this) {
            if (stopped) {
                return;
            }
            stopped = true;
        }

        // Cancel any rumble effects (may fail if device is already disconnected)
        try {
            rumble((short) 0, (short) 0);
        } catch (Exception e) {
            Log.d("DualSenseController", "Failed to cancel rumble during stop", e);
        }

        // Stop the input thread
        if (inputThread != null) {
            inputThread.interrupt();
            try {
                inputThread.join(1000);
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            }
            inputThread = null;
        }

        // Release all claimed interfaces
        if (connection != null && !ifaces.isEmpty()) {
            synchronized (ifaces) {
                for (UsbInterface iface : ifaces) {
                    try {
                        connection.releaseInterface(iface);
                    } catch (Exception e) {
                        Log.w("DualSenseController", "Failed to release interface", e);
                    }
                }
                ifaces.clear();
            }
        }

        // Close the USB connection
        if (connection != null) {
            try {
                connection.close();
            } catch (Exception e) {
                Log.w("DualSenseController", "Failed to close connection", e);
            }
        }

        // Report the device removed
        notifyDeviceRemoved();
    }

    protected void reportMotion() {
        // Report gyroscope data (in deg/s)
        notifyControllerMotion(MoonBridge.LI_MOTION_TYPE_GYRO, gyroX, gyroY, gyroZ);
        // Report accelerometer data (in m/s²)
        notifyControllerMotion(MoonBridge.LI_MOTION_TYPE_ACCEL, accelX, accelY, accelZ);
    }

    protected abstract boolean handleRead(ByteBuffer buffer);

    protected abstract boolean doInit();

    protected abstract void sendCommand(byte[] data);
}

File: app/src/main/java/com/limelight/binding/input/driver/AbstractDualSenseController.kt
package com.limelight.binding.input.driver

import android.hardware.usb.UsbConstants
import android.hardware.usb.UsbDevice
import android.hardware.usb.UsbDeviceConnection
import android.hardware.usb.UsbEndpoint
import android.hardware.usb.UsbInterface
import android.os.SystemClock
import android.util.Log

import com.limelight.nvstream.input.ControllerPacket
import com.limelight.nvstream.jni.MoonBridge

import java.nio.ByteBuffer
import java.nio.ByteOrder

abstract class AbstractDualSenseController(
    protected val device: UsbDevice,
    protected val connection: UsbDeviceConnection,
    deviceId: Int,
    listener: UsbDriverListener
) : AbstractController(deviceId, listener, device.vendorId, device.productId) {

    private var inputThread: Thread? = null
    private var stopped = false

    // Serializes output reports with the teardown sequence so no queued effect can
    // be sent after the final clear or alongside connection.close().
    protected val outputLock = Any()

    @Volatile
    protected var outputClosed = false

    protected var inEndpt: UsbEndpoint? = null
    protected var outEndpt: UsbEndpoint? = null

    // The UAC audioStreamingOut alt setting and its iso OUT endpoint, when the
    // controller exposes the DualSense audio topology. Null on non-DS5 pads.
    private var audioInterface: Pair<UsbInterface, UsbEndpoint>? = null

    // IMU data fields
    protected var gyroX = 0f
    protected var gyroY = 0f
    protected var gyroZ = 0f
    protected var accelX = 0f
    protected var accelY = 0f
    protected var accelZ = 0f

    init {
        type = MoonBridge.LI_CTYPE_PS
        capabilities = (MoonBridge.LI_CCAP_GYRO.toInt() or MoonBridge.LI_CCAP_ACCEL.toInt() or MoonBridge.LI_CCAP_RUMBLE.toInt()).toShort()
        buttonFlags = ControllerPacket.A_FLAG or ControllerPacket.B_FLAG or ControllerPacket.X_FLAG or ControllerPacket.Y_FLAG or
                ControllerPacket.UP_FLAG or ControllerPacket.DOWN_FLAG or ControllerPacket.LEFT_FLAG or ControllerPacket.RIGHT_FLAG or
                ControllerPacket.LB_FLAG or ControllerPacket.RB_FLAG or
                ControllerPacket.LS_CLK_FLAG or ControllerPacket.RS_CLK_FLAG or
                ControllerPacket.BACK_FLAG or ControllerPacket.PLAY_FLAG or ControllerPacket.SPECIAL_BUTTON_FLAG
        supportedButtonFlags = buttonFlags
    }

    private fun createInputThread(): Thread {
        return Thread {
            try {
                Thread.sleep(1000)
            } catch (e: InterruptedException) {
                return@Thread
            }

            notifyDeviceAdded()

            while (!Thread.currentThread().isInterrupted && !stopped) {
                val buffer = ByteArray(64)
                var res = -1

                do {
                    if (stopped || Thread.currentThread().isInterrupted) {
                        res = -1
                        break
                    }

                    val lastMillis = SystemClock.uptimeMillis()
                    if (inEndpt == null) {
                        Log.w("DualSenseController", "Connection or endpoint is null")
                        res = -1
                        break
                    }
                    res = connection.bulkTransfer(inEndpt, buffer, buffer.size, 3000)

                    if (res == 0) {
                        res = -1
                    }

                    if (res == -1 && SystemClock.uptimeMillis() - lastMillis < 1000) {
                        Log.d("DualSenseController", "Detected device I/O error")
                        this@AbstractDualSenseController.stop()
                        break
                    }
                } while (res == -1 && !Thread.currentThread().isInterrupted && !stopped)

                if (res == -1 || stopped || Thread.currentThread().isInterrupted) {
                    break
                }

                if (res > 0 && handleRead(ByteBuffer.wrap(buffer, 0, res).order(ByteOrder.LITTLE_ENDIAN))) {
                    reportInput()
                    reportMotion()
                }
            }
        }
    }

    private val ifaces = mutableListOf<UsbInterface>()

    override fun start(): Boolean {
        ifaces.clear()
        Log.d("DualSenseController", "start")
        for (i in 0 until device.interfaceCount) {
            val iface = device.getInterface(i)
            if (!connection.claimInterface(iface, true)) {
                Log.d("DualSenseController", "Failed to claim interface: $i")
                return false
            } else {
                ifaces.add(iface)
            }
        }
        Log.d("DualSenseController", "getInterfaceCount:" + device.interfaceCount)

        val iface = findInterface(device) ?: run {
            Log.e("DualSenseController", "Failed to find interface")
            return false
        }

        for (i in 0 until iface.endpointCount) {
            val endpt = iface.getEndpoint(i)
            if (endpt.direction == UsbConstants.USB_DIR_OUT) {
                if (outEndpt != null) {
                    Log.d("DualSenseController", "Found duplicate OUT endpoint")
                    return false
                }
                outEndpt = endpt
            } else if (endpt.direction == UsbConstants.USB_DIR_IN) {
                if (inEndpt != null) {
                    Log.d("DualSenseController", "Found duplicate IN endpoint")
                    return false
                }
                inEndpt = endpt
            }
        }
        Log.d("DualSenseController", "inEndpt: $inEndpt")
        Log.d("DualSenseController", "outEndpt: $outEndpt")

        if (inEndpt == null || outEndpt == null) {
            Log.d("DualSenseController", "Missing required endpoint")
            return false
        }

        if (!doInit()) {
            return false
        }

        discoverAudioInterface()

        inputThread = createInputThread()
        inputThread!!.start()
        return true
    }

    /**
     * Discovers the UAC audio streaming OUT interface (the alt setting that
     * carries the isochronous OUT endpoint) and notifies the listener so the
     * haptics coordinator can create a PCM pump. The pump owns the alternate
     * setting lifecycle; the interface itself was already claimed above.
     */
    private fun discoverAudioInterface() {
        for (i in 0 until device.interfaceCount) {
            val iface = device.getInterface(i)
            if (iface.interfaceClass != UsbConstants.USB_CLASS_AUDIO ||
                iface.interfaceSubclass != 0x02 // Audio Streaming
            ) {
                continue
            }
            // Alt 0 carries no endpoints; the alt 1 entry exposes the iso OUT
            // endpoint (Android surfaces each alternate setting separately).
            for (j in 0 until iface.endpointCount) {
                val ep = iface.getEndpoint(j)
                if (ep.direction == UsbConstants.USB_DIR_OUT &&
                    ep.type == UsbConstants.USB_ENDPOINT_XFER_ISOC
                ) {
                    Log.i("DualSenseController", "UAC streaming OUT iface=${iface.id} ep=0x${Integer.toHexString(ep.address)}")
                    audioInterface = iface to ep
                    listener.onDs5AudioInterfaceAvailable(deviceId, connection, iface, ep)
                    return
                }
            }
        }
    }

    override fun stop() {
        synchronized(this) {
            if (stopped) return
            stopped = true
        }

        // Tear down the haptics pump first: it needs a live connection to
        // restore the audio interface to alt 0 and release iso bandwidth.
        if (audioInterface != null) {
            audioInterface = null
            listener.onDs5AudioInterfaceGone(deviceId)
        }

        // Hold the output lock across the final clear so a report already queued by
        // the rumble worker cannot re-engage an effect after this point.
        synchronized(outputLock) {
            try {
                rumble(0, 0)
                setAdaptiveTriggers(
                    DualSenseOutputReport.BOTH_TRIGGER_FLAGS.toByte(),
                    DualSenseOutputReport.EFFECT_TYPE_OFF,
                    DualSenseOutputReport.EFFECT_TYPE_OFF,
                    ByteArray(DualSenseOutputReport.EFFECT_PAYLOAD_SIZE),
                    ByteArray(DualSenseOutputReport.EFFECT_PAYLOAD_SIZE)
                )
            } catch (e: Exception) {
                Log.d("DualSenseController", "Failed to clear controller output during stop", e)
            } finally {
                outputClosed = true
            }
        }

        inputThread?.let {
            it.interrupt()
            try {
                it.join(1000)
            } catch (e: InterruptedException) {
                Thread.currentThread().interrupt()
            }
            inputThread = null
        }

        if (ifaces.isNotEmpty()) {
            synchronized(ifaces) {
                for (iface in ifaces) {
                    try {
                        connection.releaseInterface(iface)
                    } catch (e: Exception) {
                        Log.w("DualSenseController", "Failed to release interface", e)
                    }
                }
                ifaces.clear()
            }
        }

        synchronized(outputLock) {
            try {
                connection.close()
            } catch (e: Exception) {
                Log.w("DualSenseController", "Failed to close connection", e)
            }
        }

        notifyDeviceRemoved()
    }

    protected fun reportMotion() {
        notifyControllerMotion(MoonBridge.LI_MOTION_TYPE_GYRO, gyroX, gyroY, gyroZ)
        notifyControllerMotion(MoonBridge.LI_MOTION_TYPE_ACCEL, accelX, accelY, accelZ)
    }

    protected abstract fun handleRead(buffer: ByteBuffer): Boolean

    protected abstract fun doInit(): Boolean

    protected abstract fun sendCommand(data: ByteArray)

    companion object {
        private fun findInterface(device: UsbDevice): UsbInterface? {
            for (i in 0 until device.interfaceCount) {
                val intf = device.getInterface(i)
                if (intf.interfaceClass == UsbConstants.USB_CLASS_HID && intf.endpointCount >= 2) {
                    Log.d("DualSenseController", "Found HID interface: $i")
                    return intf
                }
            }
            return null
        }
    }
}
